Sema: Fix some issues with overrides of materializeForSet

slavapestov opened the pull request:
When overriding storage with a forced static dispatch materializeForSet,
the override's materializeForSet should not override the base
materializeForSet.

This is the case where a dynamic property witnesses a protocol
requirement, and Sema synthesizes a materializeForSet for it.

In this case, the synthesized materializeForSet dynamically dispatches
to the dynamic property's getter and setter, and the protocol witness
thunk directly calls the synthesized materializeForSet.

The subclass only needs to override the getter and setter in this case,
since the base class's materializeForSet will already do the right
thing.

In fact, marking it as an override exposes a problem where we cannot
serialize an xref to an imported property's materializeForSet, since
it was not created by the importer.

Radar: <rdar://problem/35138127>

slavapestov committed:
Sema: Fix some issues with overrides of materializeForSet

When overriding storage with a forced static dispatch materializeForSet,
the override's materializeForSet should not override the base
materializeForSet.

This is the case where a dynamic property witnesses a protocol
requirement, and Sema synthesizes a materializeForSet for it.

In this case, the synthesized materializeForSet dynamically dispatches
to the dynamic property's getter and setter, and the protocol witness
thunk directly calls the synthesized materializeForSet.

The subclass only needs to override the getter and setter in this case,
since the base class's materializeForSet will already do the right
thing.

In fact, marking it as an override exposes a problem where we cannot
serialize an xref to an imported property's materializeForSet, since
it was not created by the importer.

slavapestov commented:
@swift-ci Please test

jrose-apple commented on lib/AST/ASTVerifier.cpp:
```diff
@@ -2097,10 +2097,17 @@ class Verifier : public ASTWalker {
-                 baseASD->getMaterializeForSetFunc() &&
-                 "Storage override but materializeForSet does not");
+            baseASD->isSetterAccessibleFrom(ASD->getDeclContext())) {
+          if (baseASD->getMaterializeForSetFunc()->hasForcedStaticDispatch()) {
```
For later: should this just be `final`?

jrose-apple commented on test/multifile/synthesized-accessors/materialize-for-set-1/library.swift:
```diff
@@ -0,0 +1,13 @@
+  var generatesDecimalNumbers: Bool { get set }
+}
+
+extension NumberFormatter : GeneratesDecimalNumbers {}
```
Please use the mock SDK or a handwritten class rather than the system Foundation.

jrose-apple commented on test/multifile/synthesized-accessors/materialize-for-set-1/library.swift:
```diff
@@ -0,0 +1,13 @@
+// RUN: true
```
Please use an Inputs/ subdirectory rather than a dummy RUN line.

jrose-apple commented on test/multifile/synthesized-accessors/materialize-for-set-1/main.swift:
```diff
@@ -0,0 +1,16 @@
+// RUN: %target-build-swift %S/main.swift -I %t/onone/ -emit-ir > /dev/null
+
+// RUN: %target-build-swift -emit-module -emit-module-path %t/wmo/library.swiftmodule -module-name=library -emit-library -wmo %S/library.swift
+// RUN: %target-build-swift %S/main.swift -I %t/wmo/ -emit-ir > /dev/null
```
I'm a little antsy about none of the output being checked here. Seems like this would work better as an IRGen test. There are also no calls to the variable's materializeForSet shown, either through the protocol or with a member of the class hierarchy as a static type.

slavapestov replied:
The crash was while deserializing the base class's materializeForSet, which happens when you define the override. But I can think of a way to make this more explicit.

slavapestov replied:
Yeah, instead of subclassing the subclass, the 'client' can just do an inout access on the property I guess.

jrose-apple reviewed (commented):
Fix looks good, tests need a bit of work.

slavapestov committed:
Re-organize multi-file tests a bit

- Put input files in directories named Inputs/ to eliminate bogus 'RUN: true' lines
- Remove executable_test requirement since they're not executable

slavapestov commented:
@jrose-apple I couldn't get the mock SDK working. Ran into -enable-source-import issues, and the %clang-importer-sdk-nosource variant only works with %target-swift-frontend, not %target-build-swift.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

jrose-apple commented:
Okay, but even if you have to use the real Foundation, please make your own class rather than reusing NumberFormatter.

slavapestov commented:
@jrose-apple What's the downside of using NumberFormatter (or some other Foundation class)? Are you suggesting using -import-objc-header with my own header, or something else? I actually want to build and link the result and there are multiple Swift modules in play so I'm not sure -import-objc-header makes sense either.

slavapestov commented:
@jrose-apple I updated the tests to define a new Clang module with a custom type.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@shahmishal CI is broken again?

slavapestov commented:
@swift-ci Please test

slavapestov commented:
For fuck's sake

slavapestov commented:
@swift-ci Please test

jrose-apple reviewed (approved):
Thank you. It's much better when the tests don't depend on how the SDK declares its APIs.

slavapestov commented:
@swift-ci Please test Linux

shahmishal commented:
@swift-ci Please test linux

shahmishal commented:
Its due to number of systems we have available. ![graph](https://user-images.githubusercontent.com/2727770/32083955-4f7f20ba-ba7b-11e7-8d13-d7684e840ff1.png)

https://ci.swift.org/label/mac-pro/load-statistics

slavapestov commented:
@shahmishal 

```
<unknown>:0: error: module map file '/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift-corelibs-libdispatch/dispatch/module.modulemap' not found
<unknown>:0: error: module map file '/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift-corelibs-libdispatch/dispatch/module.modulemap' not found
```

slavapestov commented:
@swift-ci Please test Linux

slavapestov commented:
```
Building remotely on ubuntu-16.04-04 (Ubuntu-16.04 linux-pr executors-4) in workspace /home/buildnode/jenkins/workspace/swift-PR-Linux
java.util.concurrent.TimeoutException: Ping started at 1509064218500 hasn't completed by 1509064458500
Caused: java.io.IOException
```

slavapestov commented:
@swift-ci Please test Linux

slavapestov commented:
@swift-ci Please test Linux

slavapestov commented:
@swift-ci Please test Linux
